Lib/fnmatch.py

```diff
@@ -77,24 +77,33 @@ def translate(pat):
     There is no way to quote meta-characters.
     """
 
-    STAR = object()
-    parts = _translate(pat, STAR, '.')
-    return _join_translated_parts(parts, STAR)
+    parts, indices = _translate(pat, '*', '.')
+    return _join_translated_parts(parts, indices)
 
+_re_setops_sub = re.compile(r'([&~|])').sub
+_re_escape = functools.lru_cache(maxsize=32768)(re.escape)
 
-def _translate(pat, STAR, QUESTION_MARK):
+def _translate(pat, star, question_mark):
     res = []
     add = res.append
+    indices = []
+
     i, n = 0, len(pat)
     while i < n:
         c = pat[i]
         i = i+1
         if c == '*':
+            # store the position of the wildcard
+            indices.append(len(res))
+            add(star)
             # compress consecutive `*` into one
-            if (not res) or res[-1] is not STAR:
-                add(STAR)
+            while i < n and pat[i] == '*':
+                i += 1
         elif c == '?':
-            add(QUESTION_MARK)
+            # Handling '?' one at a time seems to more efficient
+            # even if there are consecutive '?' that could have
+            # been written directly.
```

Comment:
This comment feels unnecessary, because this whole loop works one character at a time by default, only consuming multiple characters in a single iteration when there's a good reason (e.g. parsing a `[group]`)

Reply:
I wanted to keep this comment to avoid thinking that we could have optimized using a `while` as we did for the `*`. But we can remove this comment if you think it's too much (I'll do it tomorrow).

Reply:
Ah maybe I wasn't clear. What I wanted to do was:

```py
elif c == '?':
    add(question_mark)
	while i < n and pat[i] == '?':
        i += 1
		add(question_mark)
```

I think I tried this pattern in C and it was faster. But in Python, it didn't appear to (or perhaps it wasn't faster at all and I'm remembering incorrectly). 

I'll remove the comment since it appears to be confusing.